### ISSUE TITLE
Deferred configuration of DockerCreateContainer? How?

### ISSUE THREAD

**Issue by pskowronek** [bmuschko/gradle-docker-plugin#726]
Hi,

I'm struggling to configure DockerCreateContainer in 'deferred' manner using recent versions of gradle docker plugin. In previous versions, those which not used Properties (org.gradle.api.provider.Property) but java types it worked well by simply using doFirst - I've even seen some examples encouraging to use this way. But in versions that have Properties it seems not possible that easy, or I don't get something.

For example, let's take a look at 'binds' property (for some others it is the same). I would like to work-out paths that are gonna be bound to docker, therefore I use 'conf' task and its doFirst to do this. It seems that doFirst in DockerCreateContainer doesn't work anymore - properties of this task are already set (in configuration stage?) and the ones I set in doFirst are ignored.

The question is - how can I configure DockerCreateContainer in deferred manner?

Here is the example:
```
task conf {
  project.ext.myBind = '/wrong/path/or/not/yet/known'
  doFirst {
      // calculate binds etc (here using hardcoded value)
      project.ext.myBind = '/a/good/calculated/path'
  }
}

task myImage(type: DockerBuildImage) {
    inputDir = file('docker/myapp')
    tags.add('test/myapp:latest')
}

task myContainer(type: DockerCreateContainer, dependsOn: [conf, myImage]) { 
  targetImageId myImage.getImageId()

  doFirst {
      binds = [ "${project.ext.myBind}" : '/tmp/test' ]
  }
}
                                                                                                                                                         
task startMyContainer(type: DockerStartContainer, dependsOn: myContainer) {                                                                              
    targetContainerId myContainer.getContainerId()                                                                                                       
}
```

Gradle version: 5.0
gradle-docker-plugin version: 4.2.0

**Comment by bmuschko** [comment]
Calling `doFirst` was never a good idea in the first place, however, in the past you sometimes simply couldn't avoid it. This is different with the `Provider` concept so in all cases you should probably go with assigning a `Provider` instance if something needs to be evaluated later. A Provider isn't evaluated until you call the `get()` method which happens in a task action (execution phase). I don't think you'll need to keep your `conf` task around.

```
task myContainer(type: DockerCreateContainer, dependsOn: [conf, myImage]) { 
    targetImageId myImage.getImageId()
    binds.set(project.provider {
        def calculatedBind = <some-logic>
        [ "${calculatedBind}" : '/tmp/test' ]
    })
}
```

I am not sure if that will work right away. Didn't actually try it. It's unfortunately that `MapProperty` (will be released with Gradle 5.1) didn't exist yet in the API. Using a `Property<Map<String, String>>` feels a bit weird what that's the only way you could make it work.

**Comment by pskowronek** [comment]
Thank you for clarification and the example - it works perfectly!

_For reference for other people stamping on the same problem:_
I still need to call conf (not so simple chain of things to do before to work out all the necessary bits) so I just did something like this: 
```
task myContainer(type: DockerCreateContainer, dependsOn: [conf, myImage]) {                                                                              
  targetImageId myImage.getImageId()                                                                                                                     
                                                                                                                                                         
  binds.set(project.provider {                                                                                                                           
      println("check ${project.ext.myBind}")                                                                                                             
      [ "${project.ext.myBind}" : '/tmp/test' ]                                                                                                          
  })                                                                                                                                                     
}
```
